sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala

```diff
@@ -73,15 +94,146 @@ object HiveThriftServer2 extends Logging {
+    }
+  }
+
+
   /**
    * A inner sparkListener called in sc.stop to clean up the HiveThriftServer2
    */
-  class HiveThriftServer2Listener(val server: HiveServer2) extends SparkListener {
+  class HiveThriftServer2Listener(
+      val server: HiveServer2,
+      val conf: SparkConf) extends SparkListener {
+
     override def onApplicationEnd(applicationEnd: SparkListenerApplicationEnd): Unit = {
       server.stop()
     }
-  }
 
+    val sessionList = new mutable.HashMap[String, SessionInfo]
+    val executeList = new mutable.HashMap[String, ExecutionInfo]
+    val retainedStatements =
+      conf.getInt("spark.thriftserver.ui.retainedStatements", 200)
+    val retainedSessions =
+      conf.getInt("spark.thriftserver.ui.retainedSessions", 200)
+    var totalRunning = 0
+
+    override def onJobStart(jobStart: SparkListenerJobStart): Unit = {
+      val jobGroup = for (
+        props <- Option(jobStart.properties);
+        statement <- Option(props.getProperty(SparkContext.SPARK_JOB_GROUP_ID))
+      ) yield statement
+
+      jobGroup.map( groupId => {
+        val ret = executeList.find( _ match {
+          case (id: String, info: ExecutionInfo) => info.groupId == groupId
+        })
+        if (ret.isDefined) {
+          ret.get._2.jobId += jobStart.jobId.toString
+          ret.get._2.groupId = groupId
+        }
+      })
+    }
+
+    def onSessionCreated(ip: String, sessionId: String, userName: String = "UNKNOWN"): Unit = {
+      val info = new SessionInfo(sessionId, System.currentTimeMillis, ip, userName)
+      sessionList(sessionId) = info
+      trimSessionIfNecessary
+    }
+
+    def onSessionClosed(sessionId: String): Unit = {
+      sessionList(sessionId).finishTimestamp = System.currentTimeMillis
+    }
+
+    def onStatementStart(
+        id: String,
+        sessionId: String,
+        statement: String,
+        groupId: String,
+        userName: String = "UNKNOWN"): Unit = {
+      val info = new ExecutionInfo(statement, sessionId, System.currentTimeMillis, userName)
+      info.state = ExecutionState.STARTED
+      executeList(id) = info
+      trimExecutionIfNecessary
+      sessionList(sessionId).totalExecute += 1
+      executeList(id).groupId = groupId
+      totalRunning += 1
+    }
+
+    def onStatementParse(id: String, executePlan: String): Unit = {
+      executeList(id).executePlan = executePlan
+      executeList(id).state = ExecutionState.COMPILED
+    }
+
+    def onStatementError(id: String, errorMessage: String, errorTrace: String): Unit = {
+      executeList(id).finishTimestamp = System.currentTimeMillis
+      executeList(id).detail = errorMessage
+      executeList(id).state = ExecutionState.FAILED
+      totalRunning -= 1
+    }
+
+    def onStatementFinish(id: String): Unit = {
+      executeList(id).finishTimestamp = System.currentTimeMillis
+      executeList(id).state = ExecutionState.FINISHED
+      totalRunning -= 1
+    }
+
+    private def trimExecutionIfNecessary = synchronized {
```

Comment:
I guess it would make more sense to replace this method with something like `insertExecution`, which handles both inserting a new element and retiring the eldest element.